[Flang][OMP]Add support for DECLARE MAPPER parsing and semantics

Leporacanthicus opened the pull request:
Will hit a TODO in the lowering, which there are tests added to check for this happening.

Leporacanthicus committed:
[Flang][OMP]Add support for DECLARE MAPPER parsing and semantics

Will hit a TODO in the lowering, which there are tests added
to check for this happening.

TIFitis commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,26 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
+  BeginDeclTypeSpec();
+  PushScope(Scope::Kind::OtherConstruct, nullptr);
+  const auto &spec{std::get<parser::OmpDeclareMapperSpecifier>(x.t)};
+  if (const auto &mapperName{
```
Don't we need an else case here for when the optional name hasn't been supplied and use the "default" name?

Leporacanthicus replied:
Done!

TIFitis commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1944,6 +1947,12 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPDeclareTargetConstruct &x) {
 
+bool OmpAttributeVisitor::Pre(const parser::OpenMPDeclareMapperConstruct &x) {
+  PushContext(x.source, llvm::omp::Directive::OMPD_declare_mapper);
+
```
Nit: Remove empty line.

Leporacanthicus replied:
Done!

TIFitis reviewed (commented):
Thanks very much for the patch. I have suggested some minor changes :)

kparzysz commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3872,6 +3872,19 @@ struct OpenMPDeclareTargetConstruct {
 
+struct OmpDeclareMapperSpecifier {
+  TUPLE_CLASS_BOILERPLATE(OmpDeclareMapperSpecifier);
+  std::tuple<std::optional<Name>, TypeSpec, Name> t;
```
Parsing the type as `TypeSpec` would make it easier to print meaningful errors.  Otherwise, if the user provides an intrinsic type, the parser would just flag a syntax error, which may be confusing.

TIFitis replied:
That makes sense, thanks :)

kparzysz commented:
There are some restrictions on the DECLARE MAPPER directive that are not checked anywhere.  For example, the type can't be intrinsic or abstract.  Could you add them?

TIFitis commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,26 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
+  PushScope(Scope::Kind::OtherConstruct, nullptr);
+  const auto &spec{std::get<parser::OmpDeclareMapperSpecifier>(x.t)};
+  if (const auto &mapperName{
+          std::get<std::optional<Fortran::parser::Name>>(spec.t)}) {
```
Nit: Remove all `Fortran::`

Leporacanthicus replied:
Done!

TIFitis commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,26 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
+bool OmpVisitor::Pre(const parser::OpenMPDeclareMapperConstruct &x) {
+  AddOmpSourceRange(x.source);
+  BeginDeclTypeSpec();
+  PushScope(Scope::Kind::OtherConstruct, nullptr);
```
I think the `PushScope` should be after the `MapperName` decl, as we want the scope of the Mapper to be the _curScope_, but the `var` declared within the mapper should have a scope of just this mapper directive.

Leporacanthicus replied:
Done!

Leporacanthicus commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1058,6 +1058,7 @@ class DeclarationVisitor : public ArraySpecVisitor,
   const parser::Name *ResolveDesignator(const parser::Designator &);
   int GetVectorElementKind(
       TypeCategory category, const std::optional<parser::KindSelector> &kind);
+  std::optional<DerivedTypeSpec> ResolveDerivedType(const parser::Name &);
```
Should have "undone". This came from trying to call this function from a different place, but it's been changed. Sorry about that.

TIFitis commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3872,6 +3872,19 @@ struct OpenMPDeclareTargetConstruct {
+  TUPLE_CLASS_BOILERPLATE(OpenMPDeclareMapperConstruct);
+  CharBlock source;
+  std::tuple<Verbatim, OmpDeclareMapperSpecifier,
+      std::list<std::list<OmpMapClause>>>
```
We already have an OpenMP semantic checker in place to allow/disallow clauses. All OpenMP directives use the OMPClauseList to accept OpenMP clauses.

The llvm/include/llvm/Frontend/OpenMP/OMP.td file dictates which clauses are allowed etc. I believe for Declare Mapper we should only have map clause as a Required Clause.


This is the standard away of parsing clauses for OpenMP directives and all other directives I believe follow this.

Leporacanthicus replied:
Done!

kiranchandramohan commented on flang/lib/Parser/openmp-parsers.cpp:
```diff
@@ -836,6 +836,16 @@ TYPE_PARSER(
+TYPE_PARSER(construct<OmpDeclareMapperSpecifier>(
+    maybe(name / ":" / !":"_tok), typeSpec / "::", name))
+
+// ?.? (not 4.5) Declare Mapper Construct
```
Can you quote the relevant section in the 5.2 standard?

Leporacanthicus replied:
Done!

Leporacanthicus committed:
Review updates

Leporacanthicus commented:
I've pushed an update that fixes most of the review comments. There's one thing I forgot I'd "undone" while working out some other issues, and some of the new checks during semantics will need additional tests. This will be added shortly.

Leporacanthicus committed:
Move PushContext so mapper name is available in the outer context.

Fix mapper name creation to not "fail" with type missing by using
ConstructName detail flag.

Add tests for abstract and "not derived" type checks.

tblah commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3872,6 +3872,17 @@ struct OpenMPDeclareTargetConstruct {
   std::tuple<Verbatim, OmpDeclareTargetSpecifier> t;
 };
 
+struct OmpDeclareMapperSpecifier {
```
ultra nit: some of the declarations in this file have a comment giving the OpenMP 5.2 section number and what the parsing looks like (e.g. see DECLARE REDUCTION below). It would be nice to see one here.

Leporacanthicus replied:
Now fixed.

Leporacanthicus committed:
Update version in OMP.td

Leporacanthicus committed:
Fix test and remove redundant check

Leporacanthicus committed:
Add documentation for parsing

Leporacanthicus commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3924,7 +3937,8 @@ struct OpenMPDeclarativeConstruct {
       OpenMPDeclareSimdConstruct, OpenMPDeclareTargetConstruct,
-      OpenMPThreadprivate, OpenMPRequiresConstruct>
+      OpenMPThreadprivate, OpenMPRequiresConstruct,
+      OpenMPDeclareMapperConstruct>
```
Done!

Leporacanthicus commented on flang/lib/Parser/openmp-parsers.cpp:
```diff
@@ -836,6 +836,16 @@ TYPE_PARSER(
+TYPE_PARSER(sourced(
+    construct<OpenMPDeclareMapperConstruct>(verbatim("DECLARE MAPPER"_tok),
+        "(" >> Parser<OmpDeclareMapperSpecifier>{} / ")",
+        many("MAP" >> parenthesized(many(Parser<OmpMapClause>{}))))))
```
Done!

kiranchandramohan commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,26 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
   }
 }
 
+bool OmpVisitor::Pre(const parser::OpenMPDeclareMapperConstruct &x) {
```
I don't see a debug-dump-symbols test (`%flang_fc1 -fopenmp -fdebug-dump-symbols -o - %s 2>&1 | FileCheck %s`) or an unparse with symbols test (`! RUN: %python %S/../test_symbols.py %s %flang_fc1 -fopenmp`).

kiranchandramohan reviewed (approved):
Looks OK. But I have a few comments inline.

TIFitis commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,35 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
+    mapperName->symbol = mapperSym;
+  } else {
+    mapperSym = &MakeSymbol(
+        "default", Attrs{}, MiscDetails{MiscDetails::Kind::ConstructName});
```
I think we are missing a check here for an existing default declare mapper. There can only be one unnamed declared mapper for a certain derived type.

TIFitis replied:
Okay, that's fine then.

Leporacanthicus committed:
Add further tests, fix comments

Leporacanthicus commented on flang/test/Semantics/OpenMP/declare-mapper02.f90:
```diff
@@ -0,0 +1,10 @@
+   integer :: y
+end type t1
+
+!ERROR: ABSTRACT derived type may not be used here
```
There is a standard check as part of the "use a derived type" that checks that the type isn't an abstract type in resolve-names.cpp. I tried writing my own check earlier, but this check happens before the one in the check-omp-structure.cpp, so I think it's OK to use the existing one. Unless I'm missing something.

Leporacanthicus commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,35 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
+// This "manually" walks the tree of the cosntruct, because the order
+// elements are resolved by the normal visitor will try to resolve
+// the map clauses attached to the directive without having resolved
+// the type, so the type is figured out using the implicit rules.
```
Done (I think it's clearer)

Leporacanthicus commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,35 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
   }
 }
 
+// This "manually" walks the tree of the cosntruct, because the order
```
Done.

Leporacanthicus commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1605,6 +1609,35 @@ void OmpVisitor::Post(const parser::OpenMPBlockConstruct &x) {
+    mapperName->symbol = mapperSym;
+  } else {
+    mapperSym = &MakeSymbol(
+        "default", Attrs{}, MiscDetails{MiscDetails::Kind::ConstructName});
```
To be honest, not sure. I'm not sure how the OTHER symbol is retrieved either. 

@TIFitis Any thoughts?

TIFitis replied:
Intuitively I think it would be possible to do so.

I am working on further patches for this to have full lowering support, we can make/add changes to the parsing if needed as part of those patches I guess.

kiranchandramohan replied:
In the `if` case, we can get the name from the parse-tree node and get the symbol. But we can wait for @TIFitis's lowering patch to see whether it can be used. Either way, adding the symbol seems to have caught the duplicate `default` mapper for the same type that @TIFitis raised in another comment.

Leporacanthicus commented on flang/test/Semantics/OpenMP/declare-mapper01.f90:
```diff
@@ -0,0 +1,8 @@
+! RUN: %python %S/../test_errors.py %s %flang -fopenmp -fopenmp-version=50
+  ! Test the source code starting with omp syntax
```
Comments altered to reflect actual test intent.

Leporacanthicus commented on flang/test/Parser/OpenMP/declare-mapper-unparse.f90:
```diff
@@ -0,0 +1,28 @@
+! RUN: %flang_fc1 -fdebug-unparse-no-sema -fopenmp %s | FileCheck --ignore-case %s
```
Done!

Leporacanthicus commented:
> There are some restrictions on the DECLARE MAPPER directive that are not checked anywhere. For example, the type can't be intrinsic or abstract. Could you add them?

This is now implemented, at least for the abstract (already checked elsewhere) and "not derived" cases.

TIFitis reviewed (approved):
Thanks for the patch. LGTM :)
